src/core/ext/transport/chttp2/transport/hpack_encoder.cc

```diff
@@ -54,27 +60,182 @@
+};
 
-static grpc_slice_refcount terminal_slice_refcount(
-    grpc_slice_refcount::Type::STATIC);
-static const grpc_slice terminal_slice = {
-    &terminal_slice_refcount, /* refcount */
-    {{0, nullptr}}            /* data.refcounted */
+struct MetadataComparator {
+  typedef grpc_mdelem Type;
+  static const grpc_mdelem Null() { return {0}; }
+  static bool Equals(const grpc_mdelem md1, const grpc_mdelem md2) {
+    return md1.payload == md2.payload;
+  }
+  static void Ref(grpc_mdelem md) { GRPC_MDELEM_REF(md); }
+  static void Unref(grpc_mdelem md) { GRPC_MDELEM_UNREF(md); }
 };
 
+/* Index table management */
+template <typename H>
+static HpackEncoderIndex HpackIndex(const H* hashtable,
+                                    HpackEncoderSlotHash hash_index) {
+  return hashtable[hash_index].index;
+}
+
+template <typename R, typename H>
+static const R& GetEntry(const H* hashtable, HpackEncoderSlotHash hash_index) {
+  return hashtable[hash_index].value;
+}
+
+template <typename Cmp, typename H>
+static bool TableEmptyAt(const H* hashtable, HpackEncoderSlotHash hash_index) {
+  return Cmp::Equals(hashtable[hash_index].value, Cmp::Null());
+}
+
+template <typename Cmp, typename H, typename T>
+static bool Matches(const H* hashtable, const T& value,
+                    HpackEncoderSlotHash hash_index) {
+  return Cmp::Equals(value, hashtable[hash_index].value);
+}
+
+template <typename H>
+static void UpdateIndex(H* hashtable, HpackEncoderSlotHash hash_index,
+                        HpackEncoderIndex hpack_index) {
+  hashtable[hash_index].index = hpack_index;
+}
+
+template <typename H, typename T>
+static void SetIndex(H* hashtable, HpackEncoderSlotHash hash_index,
+                     const T& value, HpackEncoderIndex hpack_index) {
+  hashtable[hash_index].value = value;
+  UpdateIndex(hashtable, hash_index, hpack_index);
+}
+
+template <typename Cmp, typename H, typename T>
+static bool GetMatchingIndex(H* hashtable, const T& value, uint32_t value_hash,
+                             HpackEncoderIndex* index) {
+  const HpackEncoderSlotHash cuckoo_first = HASH_FRAGMENT_2(value_hash);
+  if (Matches<Cmp>(hashtable, value, cuckoo_first)) {
+    *index = HpackIndex(hashtable, cuckoo_first);
+    return true;
+  }
+#if GRPC_HPACK_ENCODER_USE_CUCKOO_HASH
+  const HpackEncoderSlotHash cuckoo_second = HASH_FRAGMENT_3(value_hash);
+
+  if (Matches<Cmp>(hashtable, value, cuckoo_second)) {
+    *index = HpackIndex(hashtable, cuckoo_second);
+    return true;
+  }
+#endif
+  return false;
+}
+
+template <typename Cmp, typename H, typename T>
+static T ReplaceOlderIndex(H* hashtable, const T& value,
+                           HpackEncoderSlotHash hash_index_a,
+                           HpackEncoderSlotHash hash_index_b,
+                           HpackEncoderIndex new_index) {
+  const uint32_t hpack_idx_a = hashtable[hash_index_a].index;
+  const uint32_t hpack_idx_b = hashtable[hash_index_b].index;
+  if (hpack_idx_a < hpack_idx_b) {
```

Comment:
this would generate a lot of dup instructions.  how about:

```
const uint32_t id = min(hpack_idx_a, hpack_idx_b);
T old = GetEntry<typename Cmp::Type>(hashtable, id);
SetIndex(hashtable, id, value, new_index);
return old;
```